sdk/go/auto/cmd.go

```diff
@@ -17,17 +17,203 @@ package auto
+	if err != nil {
+		return pulumiCommand{}, fmt.Errorf("failed to run `pulumi version`: %w", err)
+	}
+	currentVersion := strings.TrimSpace(string(out))
+	min := minimumVersion
+	if pOpts.version.GT(min) {
+		min = pOpts.version
+	}
+	version, err := parseAndValidatePulumiVersion(min, currentVersion, pOpts.skipVersionCheck)
+	if err != nil {
+		return pulumiCommand{}, err
+	}
+
+	return pulumiCommand{
+		version: version,
+		command: command,
+	}, nil
+}
+
+// InstallPulumiCommand downloads and installs the Pulumi CLI. By default the
+// CLI version matching the current SDK release is installed in
+// $HOME/.pulumi/versions/$VERSION. Use the option `PulumiCommandRoot` to
+// specify a different directory, and `PulumiCommandVersion` to install a
+// custom version.
+func InstallPulumiCommand(ctx context.Context, opts ...PulumiCommandOption) (PulumiCommand,
+	error,
+) {
+	pOpts := &pulumiCommandOptions{}
+	for _, opt := range opts {
+		opt.applyPulumiCommandOption(pOpts)
+	}
+	err := pOpts.applyDefaults()
+	if err != nil {
+		return pulumiCommand{}, err
+	}
+	fullOpts := []PulumiCommandOption{
+		PulumiCommandRoot(pOpts.root),
+		PulumiCommandVersion(pOpts.version),
+		skipVersionCheck(pOpts.skipVersionCheck),
+	}
+	pulumi, err := NewPulumiCommand(fullOpts...)
+	if err == nil {
+		// Found an installation and it satisfies the version requirement
+		return pulumi, nil
+	}
+	if runtime.GOOS == "windows" {
+		if err := installWindows(ctx, pOpts.version, pOpts.root); err != nil {
+			return pulumiCommand{}, err
+		}
+	} else {
+		if err := installPosix(ctx, pOpts.version, pOpts.root); err != nil {
+			return pulumiCommand{}, err
+		}
+	}
+	return NewPulumiCommand(fullOpts...)
+}
+
+func downloadToTmpFile(ctx context.Context, url, filePattern string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return "", err
+	}
+	resp, err := httputil.DoWithRetry(req, http.DefaultClient)
+	if err != nil {
+		return "", fmt.Errorf("failed to download %s: %w", url, err)
+	}
+	defer resp.Body.Close()
+	scriptData, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read response from %s: %w", url, err)
+	}
+	tmp, err := os.CreateTemp("", filePattern)
+	if err != nil {
+		return "", err
+	}
+	scriptPath := tmp.Name()
+	// The permissions here are ignored because the tmp file already exists.
+	// We need to explicitly call chmod below to set the desired permissions.
+	err = os.WriteFile(scriptPath, scriptData, 0o600)
+	if err != nil {
+		return "", err
+	}
+	if err = tmp.Close(); err != nil {
+		return "", err
+	}
+	err = os.Chmod(scriptPath, 0o700)
+	if err != nil {
+		return "", err
+	}
+	return scriptPath, nil
+}
+
+func installWindows(ctx context.Context, version semver.Version, root string) error {
+	scriptPath, err := downloadToTmpFile(ctx, "https://get.pulumi.com/install.ps1", "install-*.ps1")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(scriptPath)
```

Comment:
Should this come before `if err != nil {`?  Right now if we e.g. fail to write to the file for whatever reason we still leave the file behind.  It's probably not a huge deal, but it would probably still be nicer to clean it up in those cases.  Maybe something like 

```
if scriptPath != "" {
  defer os.Remove(scriptPath)
}
```

and always returning the scriptPath after writing a file succeeded might work?

Same below for the posix version.

Reply:
I think we can handle the cleanup inside `downloadToTmpFile ` and ensure we only leave the tempfile around in the success case https://github.com/pulumi/pulumi/compare/51370d3f5014188053873a974a77ff10d655a876..d8da15728cdc59a95bd49747994b6c61ae2d6cb7